impl/src/features.rs

```diff
@@ -0,0 +1,276 @@
+// and: https://github.com/illicitonion/cargo-metadata-pathologies/tree/main/platform-specific-features
+//
+// There is no plan for Cargo to expose platform data in `cargo-metadata`:
+// See https://github.com/rust-lang/cargo/pull/9982#issuecomment-965423533
+//
+// The Rust documentation says `cargo-tree` may stabilize its `--unit-graph` output, which
+// would make this function faster by avoiding mulitple invocations of `cargo-tree`.
+// https://rust-lang.github.io/rfcs/2957-cargo-features2.html#cargo-metadata
+//
+pub fn get_per_platform_features(
+  cargo_dir: &Path,
+  settings: &RazeSettings,
+  packages: &[Package],
+) -> Result<BTreeMap<PackageId, Features>> {
+  let mut triples: BTreeSet<String> = BTreeSet::new();
+  if let Some(target) = settings.target.clone() {
+    triples.insert(target);
+  }
+  if let Some(targets) = settings.targets.clone() {
+    triples.extend(targets);
+  }
+
+  let mut triple_map = BTreeMap::new();
+  for triple in triples {
+    triple_map.insert(
+      triple.clone(),
+      // TODO: This part is slow, since it runs cargo-tree per-platform.
+      run_cargo_tree(cargo_dir, triple.as_str(), packages)?,
+    );
+  }
+
+  let features: Vec<(PackageId, Features)> = transpose_keys(triple_map)
+    .into_iter()
+    .map(consolidate_features)
+    .collect();
+  let mut m = BTreeMap::new();
+  for f in features {
+    let (id, features) = f;
+    m.insert(id, features);
+  }
+  Ok(m)
+}
+
+// Runs `cargo-tree` with a very specific format argument that makes it easier
+// to extract per-platform targets.
+fn run_cargo_tree(
+  cargo_dir: &Path,
+  triple: &str,
+  packages: &[Package],
+) -> Result<BTreeMap<PackageId, BTreeSet<String>>> {
+  // TODO: remove this
+  eprintln!("Run cargo-tree for {}.", triple);
+
+  let cargo_bin: PathBuf = cargo_bin_path();
+  let mut cargo_tree = Command::new(cargo_bin);
+  cargo_tree.current_dir(cargo_dir);
+  cargo_tree
+    .arg("tree")
+    .arg("--prefix=none")
+    .arg("--frozen")
+    .arg(format!("--target={}", triple))
+    .arg("--format={p}|{f}|"); // The format to print output with
+
+  let tree_output = cargo_tree.output()?;
+  assert!(tree_output.status.success());
+
+  let text = String::from_utf8(tree_output.stdout)?;
+  let mut crates: BTreeSet<String> = BTreeSet::new();
+  for line in text.lines().filter(|line| {
+    // remove dedupe lines     // remove lines with no features
+    !(line.ends_with("(*)") || line.ends_with("||") || line.is_empty())
+  }) {
+    crates.insert(line.to_string());
+  }
+  let crate_vec: Vec<String> = crates.iter().map(|s| s.to_string()).collect();
+  make_package_map(crate_vec, packages)
+}
+
+fn make_package_map(
+  crates: Vec<String>,
+  packages: &[Package],
+) -> Result<BTreeMap<PackageId, BTreeSet<String>>> {
+  let mut package_map: BTreeMap<PackageId, BTreeSet<String>> = BTreeMap::new();
+  for c in &crates {
+    let (name, version, features) = process_line(c)?;
+    let id = find_package_id(name, version, packages)?;
```

Comment:
I definitely didn't give much though to the order of growth here, figuring it would only be a couple thousand crates at most, but maybe I haven't seen the largest use cases.

Having a lookup by package id is definitely helpful in `subplanners.rs` around line 170 here:

`      features: &self.metadata.features.get(&node.id),`

I think the fix here might be to change the data structure for `packages`, ahead of this loop.